HerPortal.BusinessLogic/ExternalServices/S3FileReader/S3FileReader.cs

```diff
@@ -0,0 +1,103 @@
+﻿using Amazon;
+using Amazon.S3;
+using Amazon.S3.Model;
+using Amazon.S3.Transfer;
+using HerPublicWebsite.BusinessLogic.ExternalServices.S3FileWriter;
+using HerPublicWebsite.BusinessLogic.Services.S3ReferralFileKeyGenerator;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace HerPortal.Data.ExternalServices.S3FileReader;
+
+public class S3FileReader : IS3FileReader
+{
+    private readonly S3FileReaderConfiguration config;
+    private readonly S3ReferralFileKeyService keyService;
+
+    private readonly ILogger logger;
+
+    public S3FileReader
+    (
+        IOptions<S3FileReaderConfiguration> options,
+        S3ReferralFileKeyService keyService,
+        ILogger<S3FileReader> logger
+    ) {
+        this.config = options.Value;
+        this.keyService = keyService;
+
+        this.logger = logger;
+    }
+    
+    public async Task<Stream> ReadFileAsync(string custodianCode, int year, int month)
+    {
+        var key = keyService.GetS3KeyFromData(custodianCode, year, month);
+
+        try
+        {
+            var s3Client = new AmazonS3Client(RegionEndpoint.GetBySystemName(config.Region));
+            var fileTransferUtility = new TransferUtility(s3Client);
+
+            return await fileTransferUtility.OpenStreamAsync(config.BucketName, key);
+        }
+        catch (AmazonS3Exception ex)
+        { 
+            logger.LogError
+            (
+                "AWS S3 error when reading CSV file from bucket: \"{BucketName}\", key: \"{Key}\". Message: \"{Message}\"",
+                config.BucketName,
+                key,
+                ex.Message
```

Comment:
Do you know if the message includes the AWS error code? IIRC at least in the node sdk it doesn't...

The error code is super useful for debugging, because you can look it up against a list - e.g. https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html#RESTErrorResponses